File: packages/web/src/algorithms/QC/ruleMixedSites.ts
import { clamp } from 'lodash'

import type { AnalysisResultWithClade, NucleotideSubstitution } from 'src/algorithms/types'
import { getQCRuleStatus } from 'src/algorithms/QC/QCRuleStatus'

export interface QCRulesConfigMixedSites {
  mixedSitesThreshold: number
}

export function ruleMixedSites(
  { nucleotideComposition }: AnalysisResultWithClade,
  _1: NucleotideSubstitution[],
  { mixedSitesThreshold }: QCRulesConfigMixedSites,
) {
  const goodBases = new Set(['A', 'C', 'G', 'T', 'N', '-'])

  const totalMixedSites = Object.keys(nucleotideComposition)
    .filter((d) => !goodBases.has(d))
    .reduce((a, b) => a + nucleotideComposition[b], 0)

  let score = 100 * (totalMixedSites / mixedSitesThreshold)
  score = clamp(score, 0, Infinity)

  const status = getQCRuleStatus(score)

  return { score, totalMixedSites, mixedSitesThreshold, status }
}

export type QCResultMixedSites = ReturnType<typeof ruleMixedSites>

File: packages/nextclade/src/qc/ruleMixedSites.cpp
#include "ruleMixedSites.h"

#include <frozen/set.h>
#include <nextclade/nextclade.h>

#include <algorithm>
#include <optional>
#include <vector>

#include "../utils/mapFind.h"
#include "getQcRuleStatus.h"

namespace Nextclade {
  namespace {
    constexpr frozen::set<Nucleotide, 6> GOOD_NUCLEOTIDES = {
      Nucleotide::A,
      Nucleotide::C,
      Nucleotide::G,
      Nucleotide::T,
      Nucleotide::N,
      Nucleotide::GAP,
    };
  }//namespace

  std::optional<QCResultMixedSites> ruleMixedSites(//
    const AnalysisResult& result,                 //
    const std::vector<NucleotideSubstitution>&,    //
    const QCRulesConfigMixedSites& config          //
  ) {
    if (!config.enabled) {
      return {};
    }

    int totalMixedSites = 0;
    for (const auto& [nuc, total] : result.nucleotideComposition) {
      if (has(GOOD_NUCLEOTIDES, nuc)) {
        continue;
      }
      totalMixedSites += total;
    }

    const auto score = std::max(0.0, 100.0 * (totalMixedSites / config.mixedSitesThreshold));
    const auto& status = getQcRuleStatus(score);

    return QCResultMixedSites{
      .score = score,
      .status = status,
      .totalMixedSites = totalMixedSites,
      .mixedSitesThreshold = config.mixedSitesThreshold,
    };
  }
}// namespace Nextclade
